### ISSUE TITLE
Permissions subsystem suggestion

### ISSUE THREAD

**Issue by ben-denham** [owncloud/core#6645]
I have been working a bit in the calendar app lately, and have noticed a fair amount of permission-logic dealing with item owners as exceptional cases. E.g. If the user owns the item, then grant them access, otherwise, check their share permissions.

Instead, would it be better to have a kind of item-permissions subsystem that covers owners as well as "sharees"?

As owners are currently defined differently for each type of item (a calendar's owner is part of a table defined within the calendar app), perhaps a hook could be implemented by apps to tell the permissions subsystem whether a given user is the owner of an item.

I have made a quick example of how the permissions subsystem might work to encapsulate existing logic:
![sequence](https://f.cloud.github.com/assets/1814227/1848270/a1b96750-7682-11e3-8c4c-804b5c6dfc80.png)

This kind of solution would not break existing permissions logic, and could provide a basis for building a more flexible permissions system (e.g. providing more detailed permissions than the standard read/write/delete/reshare/etc permissions).

TL;DR: A permissions subsystem could encapsulate permission-handling logic that is currently repeated throughout apps; standardising the way permissions are handled, and preventing unintended permission-logic errors.

I may be completely wrong about this, please let me know :)

**Comment by MorrisJobke** [comment]
I’m closing this issue because it has been inactive for a few months. As this is an enhancement and hasn’t been taken on for so long, then it seems no one has the time to implement this or there is no real need for a that complex solution.

You can also contribute directly by providing a patch – see the [developer manual](http://doc.owncloud.org/server/9.0/developer_manual/). :)

Thank you!
